### PR TITLE
Perturb SHR if bypass factor calculation appears to fail

### PR DESCRIPTION
**Pull request by rraustad**
Fully autosized coil should attempt to find a valid SHR and bypass factor during sizing. 
Addresses #4853

### CODE REVIEW

**Comment by rraustad** [comment]
@Myoldmopar @mjwitte hoping for all green lights. If not, I'll relax the tolerance slightly. Existing example files should all pass without being modified by new ValidateADP function. Those that fail, if any, will be used to find a new relaxed tolerance.

Two defect files now run with valid SHR. The defect file with hard-sized SHR can't be corrected since it's a user input. The eio for that file does show the SHR, input as 0.7, should be much lower at around 0.59.

```
Component Sizing Information, Coil:Cooling:DX:SingleSpeed, CORRIDOR_FLR_3_DXCOIL, Design Size Gross Rated Sensible Heat Ratio, 0.59604
Component Sizing Information, Coil:Cooling:DX:SingleSpeed, CORRIDOR_FLR_3_DXCOIL, User-Specified Gross Rated Sensible Heat Ratio, 0.70000
```

File: src/EnergyPlus/DXCoils.hh [hunk 1]
```diff
@@ -1404,7 +1404,7 @@ namespace DXCoils {
 	);
 
 	Real64
-	CalcCBF(
+		CalcCBF(
```
Comment (rraustad) [inline src/EnergyPlus/DXCoils.hh]: compiler auto justifies, very frustrating

File: src/EnergyPlus/ReportSizingManager.cc [hunk 2]
```diff
@@ -850,6 +855,11 @@ namespace ReportSizingManager {
 								AutosizeDes = 0.389 + 7684.0*RatedVolFlowPerRatedTotCap;
 							}
 						}
+
+						// check that the autosized SHR corresponds to a valid apperatus dew point (ADP) temperature
+						DesMassFlow = DataFlowUsedForSizing * PsyRhoAirFnPbTdbW( StdBaroPress, RatedInletAirTemp, RatedInletAirHumRat, CallingRoutine );
+						AutosizeDes = ValidateADP( CompType, CompName, RatedInletAirTemp, RatedInletAirHumRat, DataCapacityUsedForSizing, DesMassFlow, AutosizeDes, CallingRoutine );
+
```
Comment (rraustad) [inline src/EnergyPlus/ReportSizingManager.cc:862]: Right after SHR is autosized, silently test bypass factor to iterate on SHR if needed. Revised SHR is used instead and reported after this point.

File: tst/EnergyPlus/unit/DXCoils.unit.cc [hunk 3]
```diff
@@ -801,7 +801,7 @@ namespace EnergyPlus {
 		AirPressure = StdPressureSeaLevel;
 		InletAirHumRat = Psychrometrics::PsyWFnTdbTwbPb(InletDBTemp, InletWBTemp, AirPressure );
 		AirMassFlowRate = AirVolFlowRate * Psychrometrics::PsyRhoAirFnPbTdbW( AirPressure, InletDBTemp, InletAirHumRat );
-		CBF_calculated = CalcCBF( CoilType, CoilName, InletDBTemp, InletAirHumRat, TotalCap, AirMassFlowRate, SHR, AirPressure );
+		CBF_calculated = CalcCBF( CoilType, CoilName, InletDBTemp, InletAirHumRat, TotalCap, AirMassFlowRate, SHR, true, AirPressure );
```
Comment (rraustad) [inline tst/EnergyPlus/unit/DXCoils.unit.cc:804]: New flag added so CalcCBF can be ran silently without (mostly) warnings.

File: src/EnergyPlus/DXCoils.cc [hunk 4]
```diff
@@ -9992,6 +9993,110 @@ Label50: ;
...
+			OutletAirHumRat = PsyWFnTdbH( RatedInletAirTemp, HTinHumRatOut, CallingRoutine );
+			DeltaHumRat = RatedInletAirHumRat - OutletAirHumRat;
+			OutletAirEnthalpy = InletAirEnthalpy - DeltaH;
+			OutletAirTemp = PsyTdbFnHW( OutletAirEnthalpy, OutletAirHumRat );
+			OutletAirRH = PsyRhFnTdbWPb( OutletAirTemp, OutletAirHumRat, StdBaroPress, CallingRoutine );
+			if ( CBF_calculated < 1 ) {
+				CalcADPTemp = RatedInletAirTemp - ( ( RatedInletAirTemp - OutletAirTemp ) / ( 1 - CBF_calculated ) );
+				CalcADPHumRat = RatedInletAirHumRat - ( ( DeltaHumRat ) / ( 1 - CBF_calculated ) );
+				CalcADPTempFnHR = PsyTdpFnWPb( CalcADPHumRat, StdBaroPress, CallingRoutine );
+				ADPerror = CalcADPTemp - CalcADPTempFnHR;
```
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc:10071]: These 2 temps should converge when SHR, ADP, and CBF are valid.

File: tst/EnergyPlus/unit/DXCoils.unit.cc [hunk 5]
```diff
@@ -1193,4 +1193,158 @@ namespace EnergyPlus {
...
+		DXCoil( 1 ).RatedTotCap( 1 ) = 35000.0; // run right at the saturation curve
+		DXCoil( 1 ).RatedSHR( 1 ) = AutoSize;
+
+		SizeDXCoil( 1 );
+		CBF_calculated = CalcCBF( DXCoil( 1 ).DXCoilType, DXCoil( 1 ).Name, RatedInletAirTemp, RatedInletAirHumRat, DXCoil( 1 ).RatedTotCap( 1 ), DesMassFlow, DXCoil( 1 ).RatedSHR( 1 ), true );
+
+		EXPECT_NEAR( 0.67608322, DXCoil( 1 ).RatedSHR( 1 ), 0.0000001 );
+		EXPECT_NEAR( 0.0003243, CBF_calculated, 0.0000001 );
+
+		DXCoil( 1 ).RatedTotCap( 1 ) = 40000.0; // reverse perturn SHR (i.e., decrease SHR), CalcCBF would have failed with RH >= 1.0
```
Comment (rraustad) [inline tst/EnergyPlus/unit/DXCoils.unit.cc]: typo, perturb

**Comment by rraustad** [comment]
Reviewed most of the failed files. All fail with a change (increase) in coil SHR by a factor of 0.001, the value used to perturb SHR towards convergence. So this is mostly expected. What wasn't expected is that most of these files run unchanged in the debugger (i.e., ValidateADP doesn't affect results). This is strange since the only explanation for a change in SHR of 0.001 is if this new function changed the answer. There were only 2 files that were close to a tolerance of 0.01. One was at 0.009998. Another was at 0.0101. Another, _DOAS*.idf, actually failed the test miserably and the SHR had to be corrected (which it did nicely). Something must be different in the release version. Given these findings, the tolerance will be relaxed to 0.012. Code also edited to correct typo and auto justify.

File: src/EnergyPlus/DXCoils.cc [hunk 6]
```diff
@@ -9992,6 +9993,110 @@ Label50: ;
...
+			} else {
+				ADPerror = 0; // might be able to check for RH >= 1 and reduce SHR, need defect file for that since can't create one
+			}
+
+			if ( std::abs( ADPerror )  > 0.012 ) {
```
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc:10076]: Tolerance used to test convergence.

Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc:10076]: To finish this thought, the actual solution has 2 unknowns, ADP and CBF. This fix only picks the first available solution within the tolerance.

**Comment by rraustad** [comment]
Looks good.

**Comment by rraustad** [comment]
Push this up.

File: src/EnergyPlus/DXCoils.cc [hunk 7]
```diff
@@ -9850,7 +9851,7 @@ Label50: ;
 		//  Eventually inlet air conditions will be used in DX Coil, these lines are commented out and marked with this comment line
 		//  Pressure will have to be pass into this subroutine to fix this one
 		OutletAirRH = PsyRhFnTdbWPb( OutletAirTemp, OutletAirHumRat, BaroPress, RoutineName );
-		if ( OutletAirRH >= 1.0 ) {
+		if ( OutletAirRH >= 1.0 && PringFlag ) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/DXCoils.cc]: Pring :)  I can push a change if you'd like, @rraustad.

**Comment by Myoldmopar** [comment]
@mjwitte CI results look as expected.  Code changes look good.  Unit test looks good.  Any thoughts before merging?

File: src/EnergyPlus/DXCoils.cc (hunk 7 above)
Comment (rraustad) [inline src/EnergyPlus/DXCoils.cc]: Man are my eyes getting bad. I'll fix that now.

**Comment by mjwitte** [comment]
@Myoldmopar Only thing would be - did you test the defect file?  If that runs successfully now, then I have no other comments.
@rraustad This is an important fix - less frustrations for users.

**Comment by Myoldmopar** [comment]
That particular defect file fails on both develop and this branch because `KITCHEN_FLR_6_DXCOIL` has zero rated total capacity and zero rated air flow rate.  I got this with the three variations of the idfs in dev support.  I will say I didn't transition them or anything, just ran as-is.  I re-ran the one test file that shows diffs, and I can confirm the diffs shown by CI.  @rraustad did you make fixes to the defect file to get it to run?  If so, which one is the valid one to use?  Or maybe you have a local copy of the idf that runs.... let me know.  I'm happy to run to verify and get this in.

**Comment by rraustad** [comment]
I had to change the OA from 1E-17 to 0.01. I thought I pushed that up yesterday. Just pushed them up, try again using HotelLarge-Memphis-0003-MAJOR-HV02_DOASrec-v8.3.idf

```
DesignSpecification:OutdoorAir,
 SZ DSOA Kitchen_Flr_6,   !- Name
 Flow/Zone,               !- Outdoor Air Method
 ,                        !- Outdoor Air Flow per Person {m3/s-person}
 ,                        !- Outdoor Air Flow per Zone Floor Area {m3/s-m2}
 0.01;                   !- Outdoor Air Flow per Zone {m3/s}
```

**Comment by Myoldmopar** [comment]
Thanks for the updated file @rraustad.

With develop:

```
   ** Severe  ** Coil:Cooling:DX:SingleSpeed "MECH_FLR_1_DXCOIL" -- coil bypass factor calculation did not converge after max iterations.
   **   ~~~   ** The RatedSHR of [0.572], entered by the user or autosized (see *.eio file),
   **   ~~~   ** may be causing this. The line defined by the coil rated inlet air conditions
   **   ~~~   ** (26.7C drybulb and 19.4C wetbulb) and the RatedSHR (i.e., slope of the line) must intersect
   **   ~~~   ** the saturation curve of the psychrometric chart. If the RatedSHR is too low, then this
   **   ~~~   ** intersection may not occur and the coil bypass factor calculation will not converge.
   **   ~~~   ** If autosizing the SHR, recheck the design supply air humidity ratio and design supply air
   **   ~~~   ** temperature values in the Sizing:System and Sizing:Zone objects. In general, the temperatures
   **   ~~~   ** and humidity ratios specified in these two objects should be the same for each system
   **   ~~~   ** and the zones that it serves.
   **   ~~~   **  During Warmup, Environment=MEMPHIS_TN_USA HEATING 99.6% CONDITIONS, at Simulation time=01/21 00:00 - 00:10
   **  Fatal  ** Coil:Cooling:DX:SingleSpeed "MECH_FLR_1_DXCOIL" Errors found in calculating coil bypass factors
```

WIth your branch, no SHR error and file runs to completion.  

@mjwitte This is ready to merge once you are happy with CI results.

**Comment by mjwitte** [comment]
@Myoldmopar This one's all yours.

**Pull request merged**